// messages/emails.js
// const { GBP } = require('models') // introduces unresolveable circular dependency
const { GBP } = require('models/gbp')

// const { timestampToHuman } = require('utils')

const PERIOD = eval(process.env.defaultPeriod)

const timestampToHuman = (timestampSeconds) => {
  const months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']
  const days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
  return `${days[new Date(timestampSeconds * 1000).getDay() - 1]} ${new Date(timestampSeconds * 1000).getDate()} ${months[new Date(timestampSeconds * 1000).getMonth()]} ${new Date(timestampSeconds * 1000).getFullYear()}`
}

exports.customerReceipt = (fullName, merchant, repaymentNumber, paymentMethodType, paymentMethodNumber, loanID, balance, paymentTimestampSecs) => {
  const paymentDate = timestampToHuman(paymentTimestampSecs)
  const repayment = GBP.toPounds(repaymentNumber)
  const outstanding = GBP.toPounds(balance)
  const subject = 'PollenPay Repayment Receipt Available'
  const mailbody = `
    <div class="rps_4351">
      <div
        style="font-family: &quot;Open Sans&quot;, Arial, sans-serif, serif, EmojiFont; font-size: 14px; color: rgb(42, 42, 41); text-align: center; background-color: rgb(255, 255, 255);">
        <div class="x_container" style="display:inline-block; width:600px; text-align:center">
          <div>&nbsp;</div><span class="x_sg-image" style="float:none; display:block; text-align:center"><img
              data-imagetype="External"
              src="https://mk0pollenpayx1d2xtea.kinstacdn.com/wp-content/uploads/2021/03/PollenPay_Logo_RGB_Colour_P3.png"
              alt="" height="30" width="161" style="width:161px; height:30px"></span>
          <div>&nbsp;</div>
          <h3>Hi ${fullName}</h3>
          <p>The following payment has been successfully processed.</strong></p>
          <div>&nbsp;</div>
          <div
            style="display:inline-block; border-top:4px solid #fecf0b; border-bottom:1px solid #f2f2f2; border-right:1px solid #f2f2f2; border-left:1px solid #f2f2f2;border-radius:8px;-webkit-box-shadow: 3px 3px 16px 0px rgba(0,0,0,0.1);-moz-box-shadow: 3px 3px 16px 0px rgba(0,0,0,0.1);box-shadow: 3px 3px 16px 0px rgba(0,0,0,0.1);">
            <p>&nbsp;</p>
            <h3></h3>
            <div style="display:inline-block; width:220px; margin:10px; vertical-align:top"><span>Payment date</span>
              <h3>${paymentDate}</h3>
            </div>
            <div style="display:inline-block; width:220px; margin:10px; vertical-align:top"><span>Amount paid</span>
              <h3>GBP ${repayment} GBP</h3>
            </div>
            <div style="display:inline-block; width:220px; margin:10px; vertical-align:top"><span>Payment method</span>
              <h3>${paymentMethodType} ${paymentMethodNumber}</h3>
            </div>
            <div style="display:inline-block; width:220px; margin:10px; vertical-align:top"><span>Total remaining</span>
              <h3>GBP ${outstanding} GBP</h3>
            </div>
            <div style="display:inline-block; width:220px; margin:10px; vertical-align:top"><span>Merchant name</span>
              <h3>${merchant}</h3>
            </div>
            <div style="display:inline-block; width:220px; margin:10px; vertical-align:top">
              <span>PollenPay order number</span>
              <h3>${loanID}</h3>
            </div>
            <p>&nbsp;</p>
          </div>
          <p>&nbsp;</p>
          <p>Thank you for using PollenPay!</p>&nbsp;
          <h3>Need help?</h3>
          <p><a href="https://pollenpay.com/#/" target="_blank" rel="noopener noreferrer">Visit our FAQ</a></p>
          <p><span style="font-size:10px">(c) ${new Date().getFullYear()} <span data-markjs="true" class="markq5vrb6vkc" data-ogac="" data-ogab=""
                data-ogsc="" data-ogsb="" style="color: black;">PollenPay UK Ltd</span> </span></p>
        </div>
      </div>
    </div>
  `
  return {
    subject,
    mailbody
  }
}

exports.newLoanConfirmationCustomer = (merchant, repaymentNumber, paymentMethodType, startDate, paymentMethodNumber, loanID) => {
  const defaultPeriodSeconds = Math.round(PERIOD / 1000)
  const repayment = GBP.toPounds(repaymentNumber)
  const total = GBP.toPounds(repaymentNumber * 4)
  const subject = 'PollenPay Payment Receipt Available'
  const mailbody = `
    <div class="rps_4351">
      <div
        style="font-family: &quot;Open Sans&quot;, Arial, sans-serif, serif, EmojiFont; font-size: 14px; color: rgb(42, 42, 41); text-align: center; background-color: rgb(255, 255, 255);">
        <div class="x_container" style="display:inline-block; width:600px; text-align:center">
          <div>&nbsp;</div><span class="x_sg-image" style="float:none; display:block; text-align:center">
            <img data-imagetype="External"
              src="https://mk0pollenpayx1d2xtea.kinstacdn.com/wp-content/uploads/2021/03/PollenPay_Logo_RGB_Colour_P3.png"
              alt="" height="30" width="161" style="width:161px; height:30px"></span>
          <div>&nbsp;</div>
          <h3>Thank you for choosing <span data-markjs="true" class="markzr377cqww" data-ogac="" data-ogab="" data-ogsc=""
              data-ogsb="" style=" color: black;">PollenPay</span>!</h3>
          <p>Your order with <strong>${merchant}</strong> is confirmed.</p>
          <div>&nbsp;</div>
          <div style="display:inline-block; border-top:4px solid #fecf0b; border-bottom:1px solid #f2f2f2; border-right:1px solid #f2f2f2; border-left:1px solid #f2f2f2;border-radius:8px;-webkit-box-shadow: 3px 3px 16px 0px rgba(0,0,0,0.1);
      -moz-box-shadow: 3px 3px 16px 0px rgba(0,0,0,0.1);
              box-shadow: 3px 3px 16px 0px rgba(0,0,0,0.1);">
            <p>&nbsp; </p>
            <h3></h3>
            <h3>Your payment schedule</h3>
            <div
              style="display:inline-block; width:220px; background-color:#fcfcfc; border:1px solid #f3f3f3; margin:10px; border-radius:6px;">
              <h5 style="color:#fecf0b"><strong>FIRST PAYMENT</strong></h5>
              <h4>GBP ${repayment}</h4>
              <p style="color:#9a9a9b; font-size:12px">DUE DATE: ${timestampToHuman(startDate + defaultPeriodSeconds *
                0)}</p>
              <p style="color:#9a9a9b; font-size:12px">PAID </p>
            </div>
            <div
              style="display:inline-block; width:220px; background-color:#fcfcfc;; border:1px solid #f3f3f3; margin:10px; border-radius:6px;">
              <h5 style="color:#fecf0b"><strong>SECOND PAYMENT</strong></h5>
              <h4>GBP ${repayment}</h4>
              <p style="color:#9a9a9b; font-size:12px">DUE DATE: ${timestampToHuman(startDate + defaultPeriodSeconds *
                1)}</p>
              <p style="color:#9a9a9b; font-size:12px">&nbsp; </p>
            </div>
            <div
              style="display:inline-block; width:220px; background-color:#fcfcfc;; border:1px solid #f3f3f3; margin:10px; border-radius:6px;">
              <h5 style="color:#fecf0b"><strong>THIRD PAYMENT</strong></h5>
              <h4>GBP ${repayment}</h4>
              <p style="color:#9a9a9b; font-size:12px">DUE DATE: ${timestampToHuman(startDate + defaultPeriodSeconds *
                2)}</p>
              <p style="color:#9a9a9b; font-size:12px">&nbsp; </p>
            </div>
            <div
              style="display:inline-block; width:220px; background-color:#fcfcfc;; border:1px solid #f3f3f3; margin:10px; border-radius:6px;">
              <h5 style="color:#fecf0b"><strong>FOURTH PAYMENT</strong></h5>
              <h4>GBP ${repayment}</h4>
              <p style="color:#9a9a9b; font-size:12px">DUE DATE: ${timestampToHuman(startDate + defaultPeriodSeconds *
                3)}</p>
              <p style="color:#9a9a9b; font-size:12px">&nbsp; </p>
            </div>
            <p><strong>Payment method to be charged on due date:</strong></p>
            <p>${paymentMethodType} ending in ${paymentMethodNumber}</p>
            <h3></h3>
            <h3>Order details</h3>
            <div class="x_subcontainer" style="display:inline-block; width:460px; text-align:center">
              <p><span><strong>Merchant name:</strong></span><span>&nbsp;${merchant}</span></p>
              <p><span><strong>PollenPay order number:</strong></span><span>&nbsp;${loanID}</span></p>
              <h3>Total&nbsp;&nbsp;GBP ${total}</h3>
              <p>&nbsp; </p>
            </div>
          </div>
        </div>
        <p>&nbsp; </p>
        <h3>Need help?</h3>
        <p><a href="https://pollenpay.com/#/" target="_blank" rel="noopener noreferrer">Visit our FAQ</a></p>
        <p><span style="font-size:10px">(c) ${new Date().getFullYear()} <span data-markjs="true" class="markq5vrb6vkc" data-ogac="" data-ogab=""
              data-ogsc="" data-ogsb="" style="color: black;">PollenPay UK Ltd</span> </span></p>
      </div>
    </div>
  `
  return {
    subject: subject,
    mailbody: mailbody
  }
}

exports.reminder1 = (fullName, merchant, repaymentNumber, paymentMethodType, paymentMethodNumber, loanID, totalNumber, dueDate, dueIn) => {
  const paymentDate = timestampToHuman(dueDate)
  const repayment = GBP.toPounds(repaymentNumber)
  const total = GBP.toPounds(totalNumber)
  const subject = 'PollenPay Upcoming Payment Reminder'
  const mailbody = `
    <div class="rps_4351">
      <div
        style="font-family: &quot;Open Sans&quot;, Arial, sans-serif, serif, EmojiFont; font-size: 14px; color: rgb(42, 42, 41); text-align: center; background-color: rgb(255, 255, 255);">
        <div class="x_container" style="display:inline-block; width:600px; text-align:center">
          <div>&nbsp;</div><span class="x_sg-image" style="float:none; display:block; text-align:center"><img
              data-imagetype="External"
              src="https://mk0pollenpayx1d2xtea.kinstacdn.com/wp-content/uploads/2021/03/PollenPay_Logo_RGB_Colour_P3.png"
              alt="" height="30" width="161" style="width:161px; height:30px"></span>
          <div>&nbsp;</div>
          <h3>Hi ${fullName}</h3>
          <p>This is a courtesy reminder that a payment for your order from<br><strong>${merchant}</strong> is
            <strong>due ${dueIn}.</strong></p>
          <div>&nbsp;</div>
          <div style="display:inline-block; border-top:4px solid #fecf0b; border-bottom:1px solid #f2f2f2; border-right:1px solid #f2f2f2; border-left:1px solid #f2f2f2;border-radius:8px;-webkit-box-shadow: 3px 3px 16px 0px rgba(0,0,0,0.1);
    -moz-box-shadow: 3px 3px 16px 0px rgba(0,0,0,0.1);
    box-shadow: 3px 3px 16px 0px rgba(0,0,0,0.1);">
            <p>&nbsp;</p>
            <h3></h3>
            <div style="display:inline-block; width:220px; margin:10px; vertical-align:top"><span>Due date</span>
              <h3>${paymentDate} (${dueIn})</h3>
            </div>
            <div style="display:inline-block; width:220px; margin:10px; vertical-align:top"><span>Amount due</span>
              <h3>GBP${repayment}</h3>
            </div>
            <div style="display:inline-block; width:220px; margin:10px; vertical-align:top"><span>Payment method</span>
              <h3>${paymentMethodType} ${paymentMethodNumber}</h3>
            </div>
            <div style="display:inline-block; width:220px; margin:10px; vertical-align:top"><span>Order total</span>
              <h3>GBP${total}</h3>
            </div>
            <div style="display:inline-block; width:220px; margin:10px; vertical-align:top"><span>Merchant name</span>
              <h3>${merchant}</h3>
            </div>
            <div style="display:inline-block; width:220px; margin:10px; vertical-align:top"><span>PollenPay order
                number</span>
              <h3>${loanID}</h3>
            </div>
            <p>&nbsp;</p>
          </div>
          <p>&nbsp;</p>
          <p>Please make sure your default payment method is chargeable for <strong>GBP${repayment}</strong> on
            <strong>${paymentDate}</strong> to avoid a <strong>GBP 5.00</strong> late fee</p>
          <p>If you would like to pay with a different method or pay before the due date, please log in to your account.</p>
          &nbsp;
          <h3>Need help?</h3>
          <p><a href="https://pollenpay.com/#/" target="_blank" rel="noopener noreferrer">Visit our FAQ</a></p>
          <p><span style="font-size:10px">(c) 2021 <span data-markjs="true" class="markq5vrb6vkc" data-ogac="" data-ogab=""
                data-ogsc="" data-ogsb="" style="color: black;">PollenPay UK Ltd</span> </span></p>
        </div>
      </div>
    </div>
  `
  return {
    subject: subject,
    mailbody: mailbody
  }
}

exports.confirmEmail = (givenName, email, confirmationLink) => {
  const mailbody = `Hi there, ${givenName}!<br/><br/>Somebody (probably you) recently selected ${email} as their new mail address with PollenPay. We want to make sure it's you to keep you posted with important info regarding your orders with us.<br/><br/><a href="${confirmationLink}">You can follow this link to confirm your address.</a> You won't be able to make purchases until your primary is email is verified. <br/><br/>If it wasn't you disregard this email or reply to this email if you're having any issues.`
  return {
    subject: 'Verify your email with PollenPay',
    mailbody
  }
}

exports.merchantReceipt = (fullName, paymentTimestamp, itemsDescription, pollenReference, totalAmount, referenceCode, fees, amountToMerchant) => {
  const subject = `PollenPay Order ${pollenReference}`
  const date = timestampToHuman(paymentTimestamp)
  const poundFees = GBP.toPounds(fees)
  const poundsToMerchant = GBP.toPounds(amountToMerchant)
  const mailbody = `<div class="rps_4351">
  <div style="font-family: &quot;Open Sans&quot;, Arial, sans-serif, serif, EmojiFont; font-size: 14px; color: rgb(42, 42, 41); text-align: center; background-color: rgb(255, 255, 255);">
    <div class="x_container" style="display:inline-block; width:600px; text-align:center">
      <div>&nbsp;</div><span class="x_sg-image" style="float:none; display:block; text-align:center">
      <img data-imagetype="External" src="https://mk0pollenpayx1d2xtea.kinstacdn.com/wp-content/uploads/2021/03/PollenPay_Logo_RGB_Colour_P3.png" alt="" height="30" width="161" style="width:161px; height:30px"></span> <div>&nbsp;</div><h3>You have received a new purchase through <span data-markjs="true" class="markzr377cqww" data-ogac="" data-ogab="" data-ogsc="" data-ogsb="" style=" color: black;">PollenPay</span>!</h3><div>&nbsp;</div><div style="display:inline-block; border-top:4px solid #fecf0b; border-bottom:1px solid #f2f2f2; border-right:1px solid #f2f2f2; border-left:1px solid #f2f2f2;border-radius:8px;-webkit-box-shadow: 3px 3px 16px 0px rgba(0,0,0,0.1);
-moz-box-shadow: 3px 3px 16px 0px rgba(0,0,0,0.1);
        box-shadow: 3px 3px 16px 0px rgba(0,0,0,0.1);"><p>&nbsp; </p><h3>Order details</h3><div class="x_subcontainer" style="display:inline-block; width:460px; text-align:center"><p><span><strong>Customer name:</strong></span><span>&nbsp;${fullName}</span></p><p><span><strong>Date:</strong></span><span>&nbsp;${date}</span></p><p><span><strong>Description:</strong></span><span>&nbsp;${itemsDescription}</span></p><p><span><strong>Store order reference:</strong></span><p style="font-size:2.5em;"><strong><span>&nbsp;${referenceCode}</span></strong></p></p><strong>Purchase value:&nbsp;&nbsp;&#163;${totalAmount} GBP</strong><p>Less PollenPay fees:&nbsp;&nbsp;&#163;${poundFees} GBP</p><h3>You will receive:&nbsp;&nbsp;&#163;${poundsToMerchant} GBP</h3><p>&nbsp; </p></div></div>

  </div>
    <p>&nbsp; </p>
      <h3>Need help?</h3>
      <p><a href="https://pollenpay.zendesk.com/hc/en-gb/categories/900001352803-I-m-a-Customer" target="_blank" rel="noopener noreferrer">Visit our FAQ</a></p><p><span style="font-size:10px">&#169; 2022 <span data-markjs="true" class="markq5vrb6vkc" data-ogac="" data-ogab="" data-ogsc="" data-ogsb="" style="color: black;">PollenPay UK Ltd</span> </span></p>
  </div></div>`
  return {
    subject,
    mailbody
  }
}
