// messages/userErrors.js
exports.emailAlreadyInUse = {
  message: 'Email already in use! Please, input a different email.',
  errorCode: 10,
  errorDisplay: 'formErrorState'
}

exports.invalidPostalCode = {
  message: 'Invalid postal code',
  errorCode: 0,
  statusCode: 400,
  errorDisplay: 'formErrorState'
}

exports.userDoesNotExist = {
  message: 'User doesn\'t exist.',
  statusCode: 400,
  errorDisplay: 'formErrorState'
}

// same as sessionExpired, to be migrated to error v3
exports.invalidToken = {
  message: 'Invalid token',
  statusCode: 500,
  errorCode: 84,
  errorDisplay: 'errorPage'
}

exports.sessionExpired = {
  message: 'Session expired',
  statusCode: 500,
  errorCode: 84,
  errorDisplay: 'errorPage'
}

exports.wrongOTP = {
  message: 'Wrong code',
  statusCode: 400,
  errorCode: 0,
  errorDisplay: 'formErrorState'
}

exports.userBlocked = {
  message: 'Account banned',
  statusCode: 400,
  errorDisplay: 'formErrorState'
}

exports.invalidDirectCredentials = {
  message: 'Invalid credentials.',
  statusCode: 400,
  errorDisplay: 'formErrorState'
}

exports.userNotFound = {
  message: 'User not found',
  statusCode: 500,
  errorDisplay: 'errorPage'
}

exports.imageTooLarge = {
  message: 'Image must be under 10 MBs',
  statusCode: 400,
  errorCode: 10,
  errorDisplay: 'modal'
}

const PM_CAP = process.env.paymentMethodMaxCount
exports.tooManyCards = {
  message: `You may only save up to ${PM_CAP} payment methods at the same time.`,
  statusCode: 400,
  errorCode: 10,
  errorDisplay: 'modal'
}

exports.failedToAddCard = {
  message: 'We couldn\'t validate your latest added card. Please, try adding it again.',
  statusCode: 500,
  errorCode: 84,
  errorDisplay: 'errorPage'
}

exports.expirationTooSoon = {
  message: 'Your card expires in less than the maximum lenght of an BNPL purchase. Try attaching a more recently issued card, or contact your bank to renew your card.',
  statusCode: 400,
  errorCode: 10,
  errorDisplay: 'modal'
}

exports.emailNotConfirmed = {
  message: 'Email is not confirmed. Please, check your email and verify it.',
  statusCode: 500,
  errorDisplay: 'errorPage'
}

exports.invalidPhoneNumber = {
  message: 'Please, submit a valid phone number.',
  statusCode: 400,
  errorDisplay: 'modal',
  errorCode: 10
}

exports.badCard = {
  message: 'Sorry, we do not support prepaid card billing. Use a regular debit card issued by a major bank.',
  statusCode: 400,
  errorDisplay: 'modal',
  errorCode: 11
}
