// aws-layer/s3.js
const AWS = require('aws-sdk')

const BUCKET = process.env.bucket
const REGION = process.env.AWSRegion

const s3 = new AWS.S3({
  region: REGION
})

exports.putPNG = async (fileName, image) => {
  const link =  'https://' + BUCKET + '.s3.eu-west-2.amazonaws.com/' + fileName
  await s3.putObject({
    Bucket: BUCKET,
    Key: fileName, // Key attrib already encodes, don't put encoded name to avoid double encoding
    Body: image,
    ACL: 'public-read',
    ContentType: 'image/png'
  }).promise()
  return link
}

exports.putJPG = async (fileName, image) => {
  const link =  'https://' + BUCKET + '.s3.eu-west-2.amazonaws.com/' + fileName
  await s3.putObject({
    Bucket: BUCKET,
    Key: fileName, // Key attrib already encodes, don't put encoded name to avoid double encoding
    Body: image,
    ACL: 'public-read',
    ContentType: 'image/jpg'
  }).promise()
  return link
}

exports.putBLOB = async (fileName, file) => {
  const link =  'https://' + BUCKET + '.s3.eu-west-2.amazonaws.com/' + fileName
  await s3.putObject({
    Bucket: BUCKET,
    Key: file, // Key attrib already encodes, don't put encoded name to avoid double encoding
    Body: image,
    ACL: 'public-read',
  }).promise()
  return link
}
