// pp-admin/model/loanitembase.js
const { Q } = require('pp-mysql/client').shorthands();
const { SEARCH, FIELD } = require('./model');
const { LoanJSONBase } = require('./loanjsonbase');

/// Loan items base data
class LoanItemBase extends LoanJSONBase {

    static PRODUCT = {
        CTE: 'product',
        SID: 'product_sid',
        NAME: 'product_name',
        QUANTITY: 'product_quantity',
        IMAGE_URL: 'product_image_url',
        PRICE_TOTAL: 'product_price_total',
        PRICE_WOTAX: 'product_price_wotax',
        TAX_PERCENT: 'product_tax_percent',
    };

    static fieldItem = {
        [FIELD.MODE]: FIELD.MODE_JOIN,
        [FIELD.REQ]: ['loan_id'],
        [FIELD.JOIN]: 'joinItem',// function for make JOIN SQL
    };

    static fieldProduct = {
        [FIELD.MODE]: FIELD.MODE_JOIN,
        [FIELD.REQ]: [this.PRODUCT.SID],
        [FIELD.CTE]: 'cteProduct',// function for make CTE SQL
        [FIELD.JOIN]: 'joinProduct',// function for make JOIN SQL
    };

    /// API field => DB field
    static api2db = {
        ...super.api2db,
        [this.JSONORD]: {
            ...super.api2db[this.JSONORD],
            ...this.fieldItem,
        },
        [this.PRODUCT.SID]: {// Product String ID
            ...this.fieldItem,
            [FIELD.SELECT]: `${Q(this.jsontable)}.${Q(this.PRODUCT.SID)}`,
        },
        [this.PRODUCT.NAME]: {// Name of merchant product (readonly by JOIN)
            ...this.fieldProduct,
            [FIELD.SELECT]: `${Q(this.PRODUCT.CTE)}.${Q(this.PRODUCT.NAME)}`,
        },
        [this.PRODUCT.QUANTITY]: {// Quantity of merchant product (always 1)
            [FIELD.MODE]: FIELD.MODE_CALC,
            [FIELD.SELECT]: 1,// always 1 for items
        },
        [this.PRODUCT.IMAGE_URL]: {// Image url of merchant product (readonly by JOIN)
            ...this.fieldProduct,
            [FIELD.SELECT]: `${Q(this.PRODUCT.CTE)}.${Q(this.PRODUCT.IMAGE_URL)}`,
        },
        [this.PRODUCT.PRICE_TOTAL]: {// Total price of merchant product (readonly by JOIN)
            ...this.fieldProduct,
            [FIELD.SELECT]: `${Q(this.PRODUCT.CTE)}.${Q(this.PRODUCT.PRICE_TOTAL)}`,// pounds
        },
        [this.PRODUCT.PRICE_WOTAX]: {// Price without tax of merchant product (readonly by JOIN)
            ...this.fieldProduct,
            [FIELD.SELECT]: `${Q(this.PRODUCT.CTE)}.${Q(this.PRODUCT.PRICE_WOTAX)}`,// pounds
        },
        [this.PRODUCT.TAX_PERCENT]: {// Tax percent of merchant product (readonly by JOIN)
            ...this.fieldProduct,
            [FIELD.SELECT]: `${Q(this.PRODUCT.CTE)}.${Q(this.PRODUCT.TAX_PERCENT)}`,
        },
    };

    /// Search parameters
    static search = {
        [this.PRODUCT.SID]: {// Merchant product String ID
            [SEARCH.WEQ]: 200,// Weight of search word (EQ)ual field
            [SEARCH.WIEQ]: 100,// Weight of search word (I)ndepend case (EQ)ual field
            [SEARCH.WHAS]: 50,// Weight of field (HAS) search word inside
            [SEARCH.WIHAS]: 50,// Weight of field (I)ndepend case (HAS) search word inside
        },
        [this.PRODUCT.NAME]: {// Merchant product name
            [SEARCH.WEQ]: 200,// Weight of search word (EQ)ual field
            [SEARCH.WIEQ]: 100,// Weight of search word (I)ndepend case (EQ)ual field
            [SEARCH.WHAS]: 50,// Weight of field (HAS) search word inside
            [SEARCH.WIHAS]: 50,// Weight of field (I)ndepend case (HAS) search word inside
        },
    };

    /// Construct SQL for JSON table from field
    static jsonTable() {
        return `JSON_TABLE(${Q(this.table)}.${Q('items')}, '$[*]' COLUMNS(${[
            `${Q(this.JSONORD)} FOR ORDINALITY`,
            `${Q(this.PRODUCT.SID)} VARCHAR(16) path '$'`,
        ].join(',')}))`;
    }
}

module.exports = { LoanItemBase };
